qiskit/pulse/schedule.py

```diff
@@ -438,6 +515,122 @@ def _add_timeslots(self, time: int, schedule: ScheduleComponent) -> None:
                         "".format(new=schedule.name or '', old=self.name or '', time=time,
                                   ch=channel, t0=interval[0], tf=interval[1]))
 
+        _check_nonnegative_timeslot(self._timeslots)
+
+    def replace(
+            self,
+            old: ScheduleComponent,
+            new: ScheduleComponent,
+            inplace: bool = False,
+            enforce_identical_timing: bool = True,
+    ) -> 'Schedule':
+        """Return a schedule with the ``old`` instruction replaced with a ``new``
+        instruction.
+
+        The replacment matching is based on an instruction equality check.
+
+        .. jupyter-kernel:: python3
+          :id: replace
+
+        .. jupyter-execute::
+
+          from qiskit import pulse
+
+          d0 = pulse.DriveChannel(0)
+
+          sched = pulse.Schedule()
+
+          old = pulse.Play(pulse.Constant(100, 1.0), d0)
+          new = pulse.Play(pulse.Constant(100, 0.1), d0)
+
+          sched += old
+
+          sched = sched.replace(old, new)
+
+          assert sched == pulse.Schedule(new)
+
+        Only matches at the top-level of the schedule tree. If you wish to
+        perform this replacement over all instructions in the schedule tree.
+        Flatten the schedule prior to running::
+
+        .. jupyter-execute::
+
+          sched = pulse.Schedule()
+
+          sched += pulse.Schedule(old)
+
+          sched = sched.flatten()
+
+          sched = sched.replace(old, new)
+
+          assert sched == pulse.Schedule(new)
+
+        By default the replacement will fail if ``new`` and ``old`` occupy
+        different timeslots. This behavior can be disable by setting
+        ``enforce_identical_timing=False``.
+
+        .. jupyter-execute::
+
+          sched = pulse.Schedule()
+
+          old = pulse.Play(pulse.Constant(100, 1.0), d0)
+          new = pulse.Play(pulse.Constant(20, 0.1), d0)
+
+          sched += old
+
+          # Fails as new has different timing than old.
+          # sched = sched.replace(old, new)
+
+          # Turns timing check off.
+          sched.replace(old, new, inplace=True, enforce_identical_timing=False)
+
+          assert sched == pulse.Schedule(new)
+
+
+        Args:
+          old: Instruction to replace.
+          new: Instruction to replace with.
+          inplace: Replace instruction by mutably modifying this ``Schedule``.
+          enforce_identical_timing: Enforce that ``old`` and ``new``
+              occupy the same ``timeslots``.
+
+        Returns:
+          The modified schedule with ``old`` replaced by ``new``.
+
+        Raises:
+            PulseError: If the ``old`` and ``new`` pulse do not have identical
+                timeslots if ``enforce_identical_timing=True`` or if the
+                ``Schedule`` after replacement will have a timing overlap.
+        """
+        identical_timing = old.timeslots == new.timeslots
```

Comment:
Why compare `timeslots`? Why not `duration`? I can imagine wanting to replace an instruction, or wanting to replace a `CX` schedule component, but I can't think of a case where a matching channel set and total duration of the block wouldn't be sufficient

Reply:
Replacing a Schedule with another Schedule. Right now due to the way Schedule's only block their occupy timeslots it is technically possibly to insert an instruction during another Schedule on channels it occupies. While I am open to considering moving to calling a Schedule being a contiguously blocking operation on the top-level resources of the caller specified as the duration X channels of the callee, this is not the current behavior and is somewhat less flexible.

Reply:
If I remove this check would it be ok with you? With the check removed so long as the replace schedule doesn't overlap with anything else it will be valid. This was just an extra check I thought might be useful by default, but the core constraint of no overlap is still enforce by the schedule.